### DISCUSSION TITLE
How to type a decorator which also injects an attribute?

### DISCUSSION THREAD

**Discussion by BrandonNav** [python/typing#1509] [Q&A] [answered]
Hello,

I have a decorator which injects an attribute onto a function:

```python
from collections.abc import Callable
from functools import wraps
from typing import ParamSpec, TypeVar


_R = TypeVar("_R")
_P = ParamSpec("_P")


def mark_exempt(func: Callable[_P, _R]) -> Callable[_P, _R]:
    @wraps(func)
    def wrapped(*args: _P.args, **kwargs: _P.kwargs) -> _R:
        return func(*args, **kwargs)

    setattr(wrapped, "exempt", True)
    return wrapped
```

The signature of `mark_exempt` is stating that the functions takes in a callable and returns a callable of the same signature which is technically correct. But the returned function object is a different type as it has gained the new `exempt` attribute. 

I would like the type system to track the type change so that my IDE autocompletes the attribute as well as the type system being able to error if the decorator is missing.

My current attempt to solve this is as follows:

```python
from collections.abc import Callable
from functools import wraps
from typing import Literal, ParamSpec, Protocol, TypeVar, reveal_type


_R = TypeVar("_R", covariant=True)
_P = ParamSpec("_P")


class CallableExemptForCSRF(Protocol[_P, _R, _P, _R]):
    exempt: Literal[True]

    def __call__(self, *args: _P.args, **kwds: _P.kwargs) -> _R:
        ...


def mark_exempt(func: Callable[_P, _R]) -> CallableExemptForCSRF[_P, _R]:
    @wraps(func)
    def wrapped(*args: _P.args, **kwargs: _P.kwargs) -> _R:
        return func(*args, **kwargs)

    setattr(wrapped, "exempt", True)
    return wrapped  # This line errors in type checkers
```

Pyright errors due to protocol incompatibility (exempt is not present)
Mypy errors due to Incompatible return value type (got "Callable[_P, _R]", expected "CallableExemptForCSRF[_P, _R]")

If I ignore the pyright/mypy errors for `return wrapped` it does "work"

```python
@mark_exempt
def a():
    pass

reveal_type(a.exempt)
# Revealed type is "Literal[True]"
```

I don't think a # type:ignore is the way to go here. I could cast to the protocol but I tend to avoid casting as I prefer not to bypass the typechecker. An alternative is marking the protocol as runtime_checkable and adding `assert isinstance(wrapped_view, CallableExemptForCSRF)` before the return but I am unsure if that is correct (or just equivalent to the cast).

How would I type this correctly without errors?

**Comment by erictraut** [comment, marked as answer]
You're modifying the value in a way that's not type safe, so you'd need to use a `cast` call.

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBMCuB2BjALmMAbAzlAlhADmCMlAO4gCG%2BmAUKJFMgJ747wDmuBRJAwhenQUARugCmAGigAZHMjGV0UgAoVKEAMr4xiFeFSIMUgCosxANTVTEFTMhoOA%2BgCUoAXiintlkAAoARC7%2B1mAAbmo4FPDIbsYgsGIAlDSOyu5QqupaOgGp-skOiEKY2PyCIuIAogAeYgTIAGJEvBrODb7K%2BmCG6ADaqVIuALqJAFw0UJNQYrX1ozJyCgK9cQlDDlNQACZiwFCOjjaCB76YYujAUgBUauyY86kAdLeY11cA1qRb9-vKj58vRJQAC0AD59s5xptNo9YQ4aDs9hA1O9HDM6vhkL44Eh5mUhKIxP1lINnCMQeD8RUxDUMY1mq0GsTSUMoVMAALkKiYbEIRDJTaIsiUfDaLa%2BG4gO4PP4vN4AqU-J4Ku5AsEQtnQqAgMTIWAgeAwPkSuVQK4fUiAjZTM7ICjIZB%2BLmisRbKT%2BdH1YKeeJJCZTHV6g1QGx2XxUwm0%2BpNEAtNrMiFDKTOsXJIA)

```python
from functools import wraps
from typing import Callable, Literal, ParamSpec, Protocol, TypeVar, cast

_R = TypeVar("_R", covariant=True)
_P = ParamSpec("_P")

class CallableExemptForCSRF(Protocol[_P, _R]):
    exempt: Literal[True]

    def __call__(self, *args: _P.args, **kwds: _P.kwargs) -> _R:
        ...

def mark_exempt(func: Callable[_P, _R]) -> CallableExemptForCSRF[_P, _R]:
    @wraps(func)
    def wrapped(*args: _P.args, **kwargs: _P.kwargs) -> _R:
        return func(*args, **kwargs)

    setattr(wrapped, "exempt", True)
    return cast(CallableExemptForCSRF[_P, _R], wrapped)
```

**Reply by BrandonNav** [reply]
Thanks for your reply @erictraut 

So, if I'm understanding you correctly, any dynamic attribute modification is type unsafe. I imagine this is because descriptors can intercept the setattr and so you can't guarantee anything.

With this in mind and thinking about this more, since I can't modify a function in this way, I think a type-safe why to do this is using class with a `__call__`

eg:
Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGEDFjkAoUSKYAVyzTARxlK1WlwoATAqiQB9AO4gCFCkhBtw0VAE8K6LAHMhVGvXgEcOAGoEQAGigBxJFlXoY9gDLppi%2BPYAKNgQQAMoq7lAAKjpI1mosLDIASlAAvFExcQAUAETJOfZwAG426ARYqKmRIFxIAJSJ-mlQgYqh4bky-jkNCTBmFvSMZkTIAKIAHkiUqABi1HQhSbNZTi4gbgDaXfbJALp1AFwsUKdQSFMzh6bmVjabXj6Mm9W1e3vNr0gJZ1DiSMAoDIZHpvMCsjgkPBgPZuLxrgwmKMkNt-Lskgdjr9fmJJNJ5IplKoIVCYZweDAGtizpDoXgZHDSOlGT8zv9AcCYMNwbSyQAqGz6HDXLqEEBC%2Bx8vkAazkguFQP8eFl8rqUAAtAA%2BIFJLHUqAgJCoLggLBQXn0xlZAXinCSmVy229FjsqAQGzSmQXaYUVBZRkI4bMFE7HUHDXaxEjcaXX3zECLZao9F7PWnQ3G01DJExn1zBZLWbJsP%2BikNIA)


```python
from collections.abc import Callable
from functools import update_wrapper
from typing import ClassVar, Generic, Literal, ParamSpec, TypeVar


_R = TypeVar("_R", covariant=True)
_P = ParamSpec("_P")


class CallableExemptForCSRF(Generic[_P, _R]):
    exempt: ClassVar[Literal[True]] = True

    def __init__(self, func: Callable[_P, _R]):
        update_wrapper(self, func)
        self._func = func

    def __call__(self, *args: _P.args, **kwargs: _P.kwargs) -> _R:
        return self._func(*args, **kwargs)


def mark_exempt(func: Callable[_P, _R]) -> CallableExemptForCSRF[_P, _R]:
    return CallableExemptForCSRF[_P, _R](func)
```
